### ISSUE TITLE
PHP 7.2 warning deprecation

### ISSUE THREAD

**Issue by sfatfarma** [PXgamer/youtube-dl-php#11]
Hello, great code, congrats!

Starting from PHP 7.2, getting warning:

PHP Deprecated:  parse_str(): Calling parse_str() without the result argument is deprecated in \youtube-dl\Downloader.php on line 64
PHP Deprecated:  parse_str(): Calling parse_str() without the result argument is deprecated in \youtube-dl\Downloader.php on line 70

**Comment by owenvoke** [comment]
Hi @sfatfarma, thank you very much for creating an issue for this. I'll fix it in a hotfix. 👍

**Comment by sfatfarma** [comment]
Thank you!

**Comment by owenvoke** [comment]
@sfatfarma, let me know if that fixed it. Just so I can confirm. :+1:

**Comment by sfatfarma** [comment]
Hello,

Something broke in the latest version, because I am getting this error for valid videos:

Array
(
    [info_url] => 
    [vid_id] => rfOY8ePOs_0
    [info] => 
    [formats] => 
    [exploded_encodes] => 
    [status] => Array
        (
            [success] => 
            [error] => Invalid YouTube ID provided.
        )

    [rvs] => 
)


Video url is: https://www.youtube.com/watch?v=rfOY8ePOs_0

**Comment by owenvoke** [comment]
Was that working previously?

**Comment by sfatfarma** [comment]
I restored previous version, and yes, it was working before.

**Comment by sfatfarma** [comment]
In my code, I use this to get almost any YouTube ID from URLs. I know it is not pretty, but for me works:


```
                        $id = '';
                        parse_str( parse_url( $lnk, PHP_URL_QUERY ), $my_array_of_vars );
                        if(isset($my_array_of_vars['v']))
                        {
                            if(!preg_match('/^[A-Za-z0-9-_]+$/i', $my_array_of_vars['v']))
                            {
                                continue;
                            }
                            $id = $my_array_of_vars['v'];
                            
                        }
                        else
                        {
                            preg_match_all('#http(?:s)?:\/\/(?:www\.)?youtube\.com\/embed\/([^\/\s"\'?&]*)#i', $lnk, $embeds);
                            if(isset($embeds[1][0]))
                            {
                                $id = $embeds[1][0];
                            }
                            else
                            {
                                preg_match_all('#http(?:s)?:\/\/(?:www\.)?youtube\.com\/v\/([^\/\s"\'?&]*)#i', $lnk, $embeds);
                                if(isset($embeds[1][0]))
                                {
                                    $id = $embeds[1][0];
                                }
                                else
                                {
                                    preg_match_all('#http(?:s)?:\/\/(?:www\.)?youtu\.be\/([^\/\s"\'?&]*)#i', $lnk, $embeds);
                                    if(isset($embeds[1][0]))
                                    {
                                        $id = $embeds[1][0];
                                    }
                                    else
                                    {
                                        preg_match_all('#http(?:s)?:\/\/(?:www\.)?youtube\.com\/oembed\?url=http%3A\/\/(?:www\.)?youtube\.com\/watch\?v%3D([^\/\s"\'?&]*)#i', $lnk, $embeds);
                                        if(isset($embeds[1][0]))
                                        {
                                            $id = $embeds[1][0];
                                        }
                                        else
                                        {
                                            preg_match_all('#http(?:s)?:\/\/(?:www\.)?youtube\.com\/attribution_link\?a=(?:[^&])+&u=%2Fwatch%3Fv%3D([^\/\s"\'?&%]*)#i', $lnk, $embeds);
                                            if(isset($embeds[1][0]))
                                            {
                                                $id = $embeds[1][0];
                                            }
                                            else
                                            {
                                                preg_match_all('#http(?:s)?:\/\/(?:www\.)?youtube\.com\/watch\?v=([^\/\s"\'?&]*)#i', $lnk, $embeds);
                                                if(isset($embeds[1][0]))
                                                {
                                                    $id = $embeds[1][0];
                                                }
                                            }
                                        }
                                    }
                                }
                            }
                        }
```

**Comment by owenvoke** [comment]
Cheers. I'll investigate now. :+1:

**Comment by sfatfarma** [comment]
I go to sleep now, in my time zone it is late. Talk to you tomorrow. ;)

And thanks.

**Comment by owenvoke** [comment]
Looks like this was possibly due to the regex not matching, but I can't actually see where I'd have changed it... :thinking:

Could you try changing the regex to the following in `Downloader::init()`:

```php
'/(?:youtube\.com\/\S*(?:(?:\/e(?:mbed))?\/|watch\?(?:\S*?&?v\=))|youtu\.be\/)?([a-zA-Z0-9_-]{6,11})/',
```

The only thing that actually changes is adding a missing `?` to make the content of the first non-capturing group optional.

If it still doesn't work, I'll keep testing :+1:

--------------

I'm planning on a v3.0.0 release sometime in the nearish future as well, which will remove the public stuff and implement unit testing so I can prevent this happening.

This is a pretty old project that I'd half deprecated for myself, but not actually deprecated for some reason.

**Comment by sfatfarma** [comment]
Good morning, yes, using this regex things are working fine. Also, I can confirm that the PHP 7.2 warning is gone.

Thank you.

As I see it, this is a very cool project, that has it's unique features. :)

**Comment by owenvoke** [comment]
I will create another hotfix today with this change. Just out of interest, which version were you using previously? :smile:

**Comment by sfatfarma** [comment]
I was using an old version, dated from 2017.

**Comment by owenvoke** [comment]
@sfatfarma, that may have been before I updated the regex, which would make sense.

It should be fixed in v2.1.4 :+1: :tada:

If you have any further problems, open another issue and I'll look into it. :+1:

**Comment by sfatfarma** [comment]
Thank you very much.

I really enjoy working with you. If you take freelancing jobs, please let me know. :)
